### DISCUSSION TITLE
Unable to construct AdmissionRequest

### DISCUSSION THREAD

**Discussion by andlaz** [kube-rs/kube#976] [Q&A] [answered]
Hi,

i may be missing something trivial here, but i was unable to locate any sort of constructor for `AdmissionRequest` and the struct literal won't work because of the private `types` field.  The comment on this field
> /// Copied from the containing [`AdmissionReview`] and used to specify a
/// response type and version when constructing an [`AdmissionResponse`].

points to the [`TryInto<AdmissionRequest> for AdmissionReview` implementation](https://github.com/kube-rs/kube-rs/blob/fb695bc382adf72fe8f877eb3034a9bb8ee2865a/kube-core/src/admission.rs#L58) that doesn't support my use case of producing an `AdmissionReview` from scratch for unit testing a `fn (AdmissionReview) -> AdmissionReview` handler, due to it requiring the `request` field to be not `None` .. that i can't construct. Feels rather circular :-)

Any suggestions as to how to proceed from here without serde Deserializer involvement ? i can also see that the [test in admission.rs does just that](https://github.com/kube-rs/kube-rs/blob/fb695bc382adf72fe8f877eb3034a9bb8ee2865a/kube-core/src/admission.rs#L341)

thanks!
Andras

**Comment by clux** [comment, marked as answer]
Hey there! Yeah, you are right, this is quite possibly an oversight when constructing the api. We could just make it public to make unit testing easier :thinking: 

The normal  flow pattern here (`AdmissionReview` input -> `AdmissionRequest` via `TryInto` -> `AdmissionResponse` via `From`/`::invalid` -> `AdmissionReview` via `::into_review`) is not super practical for unit testing since it would just emulate what the function would do.

~~Going to move this to an issue~~ Going to open a quick PR for more review. Thanks for the question!

**Reply by clux** [reply]
Carrying on discussion in https://github.com/kube-rs/kube-rs/pull/977 :-)

**Reply by andlaz** [reply]
thank you!
